File: src/packs/holomorphic/modules/HD-09/try-session.py
"""
HD-09 Try Session: DMD with PyDMD on Skill Evolution Data

Demonstrates Dynamic Mode Decomposition on synthetic skill activation
time series. Two modes: one stable (decaying), one oscillating.

Requires: pip install numpy pydmd
"""

import numpy as np

# Note: requires pip install pydmd
# from pydmd import DMD


def run_try_session():
    """Demonstrate DMD on synthetic skill activation data."""
    # Create synthetic skill activation time series
    # Two measurement channels over 100 time steps
    t = np.linspace(0, 10, 100)

    # Mode 1: decaying oscillation (inside unit circle)
    x1 = np.exp(-0.1 * t) * np.cos(2 * np.pi * 0.5 * t)

    # Mode 2: slowly decaying oscillation (closer to unit circle)
    x2 = np.exp(-0.05 * t) * np.sin(2 * np.pi * 0.3 * t)

    # Stack into snapshot matrix (rows = channels, columns = time)
    data = np.vstack([x1, x2])

    print("=== HD-09 Try Session: DMD with PyDMD ===\n")
    print(f"Snapshot matrix shape: {data.shape}")
    print(f"  Rows (measurement channels): {data.shape[0]}")
    print(f"  Columns (time snapshots):    {data.shape[1]}")
    print(f"Time span: {t[0]:.1f} to {t[-1]:.1f}")
    print(f"Number of snapshots: {len(t)}")

    # If PyDMD is available, uncomment:
    # dmd = DMD(svd_rank=2)
    # dmd.fit(data)
    # print(f"\nDMD eigenvalues: {dmd.eigs}")
    # print(f"DMD modes shape: {dmd.modes.shape}")
    # for i, eig in enumerate(dmd.eigs):
    #     mag = abs(eig)
    #     label = "decaying" if mag < 1 else "growing" if mag > 1 else "neutral"
    #     print(f"  lambda_{i} = {eig:.4f}  |lambda| = {mag:.4f}  ({label})")

    print("\n--- Expected behavior ---")
    print("  Mode 1 (decay=0.1): eigenvalue inside unit circle (attracting)")
    print("  Mode 2 (decay=0.05): eigenvalue closer to unit circle (slower decay)")
    print("  Both modes are oscillatory (nonzero imaginary part)")


if __name__ == "__main__":
    run_try_session()

File: src/packs/holomorphic/modules/HD-09/try-session.ts
import { magnitude, argument } from '../../index.js';
import type { ComplexNumber } from '../../types.js';

/**
 * HD-09 Try Session: Dynamic Mode Decomposition
 *
 * Demonstrates DMD eigenvalue classification by manually constructing
 * eigenvalues and showing how their position relative to the unit circle
 * determines the behavior of the corresponding dynamic mode.
 */
export function runTrySession(): void {
  // Sample eigenvalues representing different dynamic behaviors
  const eigenvalues: ComplexNumber[] = [
    { re: 0.95, im: 0.1 },   // decaying oscillation (inside unit circle)
    { re: 1.0, im: 0.0 },    // persistent / neutral (on unit circle)
    { re: -0.5, im: 0.0 },   // decaying, non-oscillatory (inside, real)
    { re: 1.1, im: 0.2 },    // growing oscillation (outside unit circle)
    { re: 0.0, im: 0.99 },   // near-neutral oscillation (inside, pure imaginary)
  ];

  console.log('=== HD-09: DMD Eigenvalue Classification ===\n');
  console.log('Unit circle rule: |lambda| < 1 = decaying, |lambda| = 1 = persistent, |lambda| > 1 = growing\n');

  for (const eig of eigenvalues) {
    const mag = magnitude(eig);
    const angle = argument(eig);
    const isOscillatory = Math.abs(angle) > 0.1;

    let type: string;
    if (mag < 0.999) {
      type = isOscillatory ? 'oscillating decay' : 'decaying (attracting)';
    } else if (mag > 1.001) {
      type = isOscillatory ? 'oscillating growth' : 'growing (repelling)';
    } else {
      type = 'persistent (neutral)';
    }

    const sign = eig.im >= 0 ? '+' : '-';
    console.log(
      `  lambda = ${eig.re.toFixed(2)} ${sign} ${Math.abs(eig.im).toFixed(2)}i` +
      `  |  |lambda| = ${mag.toFixed(3)}` +
      `  |  angle = ${(angle * 180 / Math.PI).toFixed(1)} deg` +
      `  |  ${type}`
    );
  }

  console.log('\n--- Holomorphic dynamics bridge ---');
  console.log('  Inside unit circle  =>  attracting fixed point (multiplier |lambda| < 1)');
  console.log('  On unit circle      =>  periodic orbit / neutral point');
  console.log('  Outside unit circle =>  repelling fixed point (multiplier |lambda| > 1)');
}
